include/contraction/linearContraction.hpp

```diff
@@ -50,149 +50,101 @@ namespace contraction {
-         adjacent_vertices.pop_front();
-
-         pgassert(v != u);
-         pgassert(v != w);
-         pgassert(u != w);
-
-         if (graph.is_directed()) {
-             /*
-              *  u --> v --> w
-              */
-             process_shortcut(graph, u, v, w);
-             /*
-              *  w --> v --> u
-              */
-             process_shortcut(graph, w, v, u);
-
-         } else {
-             pgassert(graph.is_undirected());
-             /*
-              * u - v - w
-              */
-             process_shortcut(graph, u, v, w);
-         }
-
-         graph[v].contracted_vertices().clear();
-         boost::clear_vertex(v, graph.graph);
-         m_linearVertices -= v;
-
-         if (is_contractible(graph, u)) {
-             one_cycle(graph, u);
-         } else {
-             m_linearVertices -= u;
-         }
-         if (is_contractible(graph, w)) {
-             one_cycle(graph, w);
-         } else {
-             m_linearVertices -= w;
-         }
-     }
-
-
-
-     /**
-      *
-      * u ----e1{v1}----> v ----e2{v2}----> w
-      *
-      * e1: min cost edge from u to v
-      * e2: min cost edge from v to w
-      *
-      * result:
-      * u ---{v+v1+v2}---> w
-      *
-      */
-     void process_shortcut(G &graph, V u, V v, V w) {
-         auto e1 = graph.get_min_cost_edge(u, v);
-         auto e2 = graph.get_min_cost_edge(v, w);
-
-         if (std::get<2>(e1) && std::get<2>(e2)) {
-             auto contracted_vertices = std::get<1>(e1) + std::get<1>(e2);
-             double cost = std::get<0>(e1) + std::get<0>(e2);
-             contracted_vertices += graph[v].id;
-             contracted_vertices += graph[v].contracted_vertices();
-
-             // Create shortcut
-             CH_edge shortcut(
-                     get_next_id(),
-                     graph[u].id,
-                     graph[w].id,
-                     cost);
-             shortcut.contracted_vertices() = contracted_vertices;
-
-             graph.add_shortcut(shortcut, u, w);
-         }
-     }
-
-
- private:
-     Identifiers<V> m_linearVertices;
-     Identifiers<V> m_forbiddenVertices;
-
-     int64_t last_edge_id;
+    typedef typename G::V V;
+    typedef typename G::V_i V_i;
+    typedef typename G::B_G B_G;
+
+    Identifiers<V> m_linearVertices;
+    Identifiers<V> m_forbiddenVertices;
+    int64_t last_edge_id;
+
+public:
+    void operator()(G &graph, Identifiers<V>& forbidden_vertices) {
+        doContraction(graph, forbidden_vertices);
+    }
+
+    Pgr_linear():last_edge_id(0) {}
+
+    void setForbiddenVertices(
+            Identifiers<V> forbidden_vertices) {
+        m_forbiddenVertices = forbidden_vertices;
+    }
+
+    bool is_contractible(G &graph, V v) {
+        return graph.is_linear(v) && !m_forbiddenVertices.has(v);
+    }
+
+    void calculateVertices(G &graph) {
+        m_linearVertices.clear();
+        V_i vi;
+        BGL_FORALL_VERTICES_T(v, graph.graph, B_G) {
```

Comment:
could be replaced by standard for loop?

Reply:
same as before, it seems to be the standard way